app/scripts/background.js

```diff
@@ -378,14 +382,23 @@ function setupController (initState, initLangCode) {
 
       if (processName === ENVIRONMENT_TYPE_FULLSCREEN) {
         const tabId = remotePort.sender.tab.id
-        openMetamaskTabsIDs[tabId] = true
+        openMetamaskTabsIDs[tabId] = { opener: previousTabId || null }
 
         endOfStream(portStream, () => {
           delete openMetamaskTabsIDs[tabId]
           controller.isClientOpen = isClientOpenStatus()
         })
       }
     } else {
+      if (remotePort.sender && remotePort.sender.tab) {
+        const tabId = remotePort.sender.tab.id
+
+        remotePort.onMessage.addListener(msg => {
+          if (msg.data && msg.data.method === 'eth_requestAccounts') {
+            previousTabId = tabId
```

Comment:
Grabbing `tab.id` from the `sender` here is a fantastic idea.

I don't feel _great_ about inferring the opener from the last tab to send `eth_requestAccounts`. There's no guarantee that's true. For example, what if `eth_requestAccounts` is sent twice in quick succession? Then we'd have a race condition, where it'd be wrong some of the time.

Another problem is that this'll get set every time a fullscreen tab opens. So you could click "Expand View" in the popup, and it'd think it was opened by whatever dapp last sent `eth_requestAccounts`.

I'm not sure I have a better alternative though 🤔. Maybe this will have to be good enough for now. Though I would at least suggest renaming this variable to make it more clear what it can and cannot be depended on for. Something like... `lastRequestedAccountsTabId`? As it's not really the "previous tab", and it'd be dangerous to treat it that way.

Reply:
Maybe this could be improved by grabbing the sender URL as well, and getting the origin from that. Then the `tabId` could be stored _under_ the origin it was sent by.
e.g. something like...

```
if (remotePort.sender && remotePort.sender.tab && remotePort.sender.url) {
  const tabId = remotePort.sender.tab.id
  const url = new URL(remotePort.sender.url)
  const origin = url.origin

  remotePort.onMessage.addListener(msg => {
    if (msg.data && msg.data.method === 'eth_requestAccounts') {
      requestAccountTabIds[origin] = tabId
    }
  })
}
```

Then on the connect screen, we could lookup the origin in the request and find the `tabId` of the last tab to send `eth_requestAccounts` for that origin.

This _technically_ doesn't solve the race condition case fully, as we _might_ have successive `eth_requestAccounts` calls from different tabs from the same origin. But... especially since it's scoped by origin, I don't really think this is a problem.

This solves the second problem as well, as we're no longer assigning openers to tabs _at all_, which is good (because we don't reliably know).Instead we're caching the specific information we need, which is who called `eth_requestAccounts` last for that origin.

Reply:
Done in 611410a1d